packages/flutter/lib/src/material/adaptive.dart

```diff
@@ -0,0 +1,58 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/widgets.dart';
+
+import 'theme.dart';
+
+/// Options for the adaptive behavior of [AdaptiveWidget].
+enum AdaptiveMode {
+  /// Always adapt the widget to the current [TargetPlatform].
+  adaptive,
+  /// Never adapt the widget to the current [TargetPlatform]. Always return
+  /// the Material Design widget.
+  static,
+  /// Adapt the widget to the current [TargetPlatform] if the widget type is
+  /// enabled in [ThemeData.adaptiveWidgetTheme].
+  theme,
+}
+
+/// A [StatelessWidget] that can build different widgets according to the
+/// current platform environment depending on the [adaptiveMode].
+abstract class AdaptiveWidget extends StatelessWidget {
```

Comment:
this name might be a little overly general for something that will rarely be used by anyone directly. How about something like "MaterialCupertinoBuilder" or something (see also comments below which will make this name make a bit more sense)

Reply:
How's PlatformBuilder? Since that's what we're ultimately switching on. The Material and Cupertino parts are more usage patterns than structurally part of the API.

Reply:
PlatformBuilder might work.